### PR TITLE
Simplify code

### PR DESCRIPTION
**Pull request by andriikushch**
This PR contains simplification to the state machine used in the code.

1. It makes `fsmS` testable by introducing `fsmAgent` interface, outgoing connection can be mocked.
2. FSM now uses its own logger.
3. Retries functions/delays are configurable (can be injected), which simplifies testing a lot.
4. Callbacks are removed from the `fsmS`.

### CODE REVIEW

File: fsm.go [hunk 1]
```diff
@@ -28,20 +28,38 @@ const (
...
+func newFSM(agent fsmAgent, logger LeveledLogger) *fsmS {
+	logger.Warn("Stan is on the scene. Starting Instana instrumentation.")
+	logger.Debug("initializing fsm")
 
 	ret := &fsmS{
-		agent:       agent,
-		retriesLeft: maximumRetries,
+		agent:                      agent,
+		retriesLeft:                maximumRetries,
+		expDelayFunc:               expDelay,
```
Comment (sanojsubran) [inline fsm.go:60]: Query : Here I was looking from where `expDelayFunc` was taking value. Does it make sense to move the `expDelay` to the top of the file. Or is it preferred to the keep the constructor function at the top?

File: agent.go [hunk 2]
```diff
@@ -360,6 +360,14 @@ func (agent *agentS) setHost(host string) {
 	agent.host = host
 }
 
+func (agent *agentS) log() LeveledLogger {
```
Comment (sanojsubran) [inline agent.go]: Minor: Inconsistency in the names of the getter functions

**Review by willianpc — APPROVED** [review]
LGTM.
I don't have further comments.

File: agent.go (hunk 2 above)
Comment (andriikushch) [reply agent.go]: I think this function was already removed in the following commits, wasn't it?

File: fsm.go (hunk 1 above)
Comment (andriikushch) [reply fsm.go:60]: `expDelay` is used in multiple places, it doesn't belong to this particular type. It is also used in other places.

**Pull request merged**